src/backend/distributed/planner/distributed_planner.c

```diff
@@ -597,31 +576,62 @@ CreateDistributedSelectPlan(uint64 planId, Query *originalQuery, Query *query,
 	MultiTreeRoot *logicalPlan = NULL;
 	List *subPlanList = NIL;
 
-	/*
-	 * For select queries we, if router executor is enabled, first try to
-	 * plan the query as a router query. If not supported, otherwise try
-	 * the full blown plan/optimize/physical planing process needed to
-	 * produce distributed query plans.
-	 */
-
-	distributedPlan = CreateRouterPlan(originalQuery, query,
-									   plannerRestrictionContext);
-	if (distributedPlan != NULL)
+	if (IsModifyCommand(originalQuery))
 	{
+		EnsureModificationsCanRun();
+
+		if (InsertSelectIntoDistributedTable(originalQuery))
+		{
+			distributedPlan =
+				CreateInsertSelectPlan(originalQuery, plannerRestrictionContext);
+		}
+		else
+		{
+			/* modifications are always routed through the same planner/executor */
+			distributedPlan =
+				CreateModifyPlan(originalQuery, query, plannerRestrictionContext);
+		}
+
+		/* the functions above always return a plan, possibly with an error */
+		Assert(distributedPlan);
+
 		if (distributedPlan->planningError == NULL)
 		{
-			/* successfully created a router plan */
 			return distributedPlan;
 		}
 		else
```

Comment:
> Open question: Do you see a simple way to not try recursive planning if we're sure that the query wouldn't be planned at all?

We're not doing anything special, just checking for CTEs and subqueries, which is pretty cheap and queries that don't work usually don't become a peformance bottleneck. The fact that we have a DEBUG message and an ERROR message is just an artifact from going through `RaiseDeferredError(distributedPlan->planningError, DEBUG1);`. 

We could consider immediately throwing an error if the problem cannot be resolved through recursive planning or filling in parameters.

> Another difference would be in the use of volatile functions, we'd be providing some workarounds. Do you see any problems with that?:

Not really, I guess we could allow the `value_2 * random()` in case of RF=1 (though not easily), but we can't do any workaround because it needs to be evaluated for every row. Conversely `(SELECT random())` resolves to the same value for all rows, so recursively planning it is correct.